plugins/field-dependent-dropdown/src/field_dependent_dropdown.ts

```diff
@@ -0,0 +1,354 @@
+/**
+ * @license
+ * Copyright 2023 Google LLC
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+/**
+ * @fileoverview A tool for making a dropdown field's options depend on the
+ * value of another field.
+ */
+
+import * as Blockly from 'blockly/core';
+import {Block} from 'blockly/core/block';
+import {FieldDropdown, MenuOption, MenuGeneratorFunction} from 'blockly/core/field_dropdown';
+import {FieldValidator} from 'blockly/core/field';
+
+/**
+ * A deep equality comparison between the two provided arrays, recursively
+ * comparing any child elements that are also arrays.
+ *
+ * @param a The first array to compare.
+ * @param b The second array to compare.
+ * @returns Whether the arrays are deeply equivalent.
+ */
+function arraysAreEquivalent<T>(a: T[], b: T[]): boolean {
+  return a.length === b.length && a.every((aElement, index) => {
+    const bElement = b[index];
+    if (Array.isArray(aElement) && Array.isArray(bElement)) {
+      return arraysAreEquivalent(aElement, bElement);
+    }
+    return aElement === bElement;
+  });
+}
+
+/** The structure of a serialized DependentDropdownOptionsChange. */
+export interface DependentDropdownOptionsChangeJson
+    extends Blockly.Events.BlockBaseJson {
+  name: string;
+  newValue: string;
+  oldValue: string;
+  oldOptions: MenuOption[];
+  newOptions: MenuOption[];
+}
+
+/** The type of the mapping from parent value to child options. */
+export interface ChildOptionMapping{
+  [key: string]: MenuOption[];
+}
+
+declare module 'blockly/core/field_dropdown' {
+  // Add a new property to dropdown fields for recording changes to options.
+  interface FieldDropdown {
+    dependentOptions?: MenuOption[];
+  }
+}
+
+/**
+ * A change event representing a simultaneous change to a dropdown field's
+ * options and value. The old value must be one of the old options, and the new
+ * value must be one of the new options. However, unlike a normal value change
+ * event, it's possible for this event to change the value to something that
+ * wasn't previously one of the valid options, in either direction, by also
+ * changing the options at the same time.
+ */
+class DependentDropdownOptionsChange extends Blockly.Events.BlockBase {
```

Comment:
Can you move the event stuff to its own file?

Reply:
Done.